From af040edbef041f99566501538f7b9fb7536fe102 Mon Sep 17 00:00:00 2001
From: Patrick Delaunay <patrick.delaunay@foss.st.com>
Date: Mon, 5 Feb 2024 19:02:59 +0100
Subject: [PATCH 880/954] feat(stm32mp2): add DDR power OFF function

Add a function stm32mp_board_ddr_power_off to power off the DDR safely
only used in BL31.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git c23ee6a1faae796d8c9739bb2a63d18654f63052 feat(stm32mp2): add DDR power OFF function]
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Change-Id: I20002010f320819dfa8047d3be9a4d90bd307c79
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/355240
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
Tested-by: Maxime MERE <maxime.mere@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Maxime MERE <maxime.mere@foss.st.com>
---
 include/drivers/st/stm32mp_ddr.h |   1 +
 plat/st/stm32mp2/plat_ddr.c      | 106 +++++++++++++++++++++++++++++++
 2 files changed, 107 insertions(+)

diff --git a/include/drivers/st/stm32mp_ddr.h b/include/drivers/st/stm32mp_ddr.h
index 970ff19e84..aca5f6144a 100644
--- a/include/drivers/st/stm32mp_ddr.h
+++ b/include/drivers/st/stm32mp_ddr.h
@@ -80,5 +80,6 @@ void stm32mp_ddr_set_qd3_update_conditions(struct stm32mp_ddrctl *ctl);
 void stm32mp_ddr_unset_qd3_update_conditions(struct stm32mp_ddrctl *ctl);
 void stm32mp_ddr_wait_refresh_update_done_ack(struct stm32mp_ddrctl *ctl);
 int stm32mp_board_ddr_power_init(enum ddr_type ddr_type);
+int stm32mp_board_ddr_power_off(void);
 
 #endif /* STM32MP_DDR_H */
diff --git a/plat/st/stm32mp2/plat_ddr.c b/plat/st/stm32mp2/plat_ddr.c
index ffd7aca96c..818193402c 100644
--- a/plat/st/stm32mp2/plat_ddr.c
+++ b/plat/st/stm32mp2/plat_ddr.c
@@ -12,7 +12,10 @@
 #include <common/fdt_wrappers.h>
 #include <drivers/delay_timer.h>
 #include <drivers/st/regulator.h>
+#include <drivers/st/regulator_fixed.h>
+#include <drivers/st/stm32mp2_ddr_helpers.h>
 #include <drivers/st/stm32mp_ddr.h>
+#include <drivers/st/stm32mp_pmic.h>
 
 #include <platform_def.h>
 
@@ -30,6 +33,7 @@ static void ddr3_supply_read(void *fdt, int node, struct ddr3_supply *supply)
 	supply->vtt = regulator_get_by_supply_name(fdt, node, "vtt");
 }
 
+#if defined(IMAGE_BL2)
 static int ddr_power_init(void *fdt, int node)
 {
 	int status;
@@ -61,6 +65,27 @@ static int ddr_power_init(void *fdt, int node)
 
 	return regulator_enable(supply.vdd);
 }
+#endif /* IMAGE_BL2 */
+
+#if defined(IMAGE_BL31)
+static void ddr_power_off(void *fdt, int node)
+{
+	struct ddr3_supply supply;
+
+	ddr3_supply_read(fdt, node, &supply);
+
+	if (supply.vtt != NULL) {
+		regulator_disable(supply.vtt);
+	}
+	if (supply.vref != NULL) {
+		regulator_disable(supply.vref);
+	}
+	if (supply.vdd != NULL) {
+		udelay(1000);
+		regulator_disable(supply.vdd);
+	}
+}
+#endif /* IMAGE_BL31 */
 #endif /* STM32MP_DDR3_TYPE */
 
 #if STM32MP_DDR4_TYPE
@@ -79,6 +104,7 @@ static void ddr4_supply_read(void *fdt, int node, struct ddr4_supply *supply)
 	supply->vtt = regulator_get_by_supply_name(fdt, node, "vtt");
 }
 
+#if defined(IMAGE_BL2)
 static int ddr_power_init(void *fdt, int node)
 {
 	int status;
@@ -121,6 +147,30 @@ static int ddr_power_init(void *fdt, int node)
 
 	return regulator_enable(supply.vdd);
 }
+#endif /* IMAGE_BL2 */
+
+#if defined(IMAGE_BL31)
+static void ddr_power_off(void *fdt, int node)
+{
+	struct ddr4_supply supply;
+
+	ddr4_supply_read(fdt, node, &supply);
+
+	if (supply.vtt != NULL) {
+		regulator_disable(supply.vtt);
+	}
+	if (supply.vref != NULL) {
+		regulator_disable(supply.vref);
+	}
+	if (supply.vdd != NULL) {
+		regulator_disable(supply.vdd);
+	}
+	if (supply.vpp != NULL) {
+		udelay(1000);
+		regulator_disable(supply.vpp);
+	}
+}
+#endif /* IMAGE_BL31 */
 #endif /* STM32MP_DDR4_TYPE */
 
 #if STM32MP_LPDDR4_TYPE
@@ -137,6 +187,7 @@ static void lpddr4_supply_read(void *fdt, int node, struct lpddr4_supply *supply
 	supply->vddq = regulator_get_by_supply_name(fdt, node, "vddq");
 }
 
+#if defined(IMAGE_BL2)
 static int ddr_power_init(void *fdt, int node)
 {
 	int status;
@@ -180,8 +231,30 @@ static int ddr_power_init(void *fdt, int node)
 
 	return regulator_enable(supply.vddq);
 }
+#endif /* IMAGE_BL2 */
+
+#if defined(IMAGE_BL31)
+static void ddr_power_off(void *fdt, int node)
+{
+	struct lpddr4_supply supply;
+
+	lpddr4_supply_read(fdt, node, &supply);
+
+	if (supply.vddq != NULL) {
+		regulator_disable(supply.vddq);
+	}
+	if (supply.vdd2 != NULL) {
+		regulator_disable(supply.vdd2);
+	}
+	if (supply.vdd1 != NULL) {
+		udelay(1000);
+		regulator_disable(supply.vdd1);
+	}
+}
+#endif /* IMAGE_BL31 */
 #endif /* STM32MP_LPDDR4_TYPE */
 
+#if defined(IMAGE_BL2)
 int stm32mp_board_ddr_power_init(enum ddr_type ddr_type)
 {
 	void *fdt = NULL;
@@ -212,3 +285,36 @@ int stm32mp_board_ddr_power_init(enum ddr_type ddr_type)
 
 	return ddr_power_init(fdt, node);
 }
+#endif /* IMAGE_BL2 */
+
+#if defined(IMAGE_BL31)
+int stm32mp_board_ddr_power_off(void)
+{
+	void *fdt = NULL;
+	int node;
+
+	if (fdt_get_address(&fdt) == 0) {
+		ddr_sub_system_clk_off();
+		return -FDT_ERR_NOTFOUND;
+	}
+
+	node = fdt_node_offset_by_compatible(fdt, -1, DT_DDR_COMPAT);
+	if (node < 0) {
+		ddr_sub_system_clk_off();
+		ERROR("%s: Cannot read DDR node in DT\n", __func__);
+		return -EINVAL;
+	}
+
+	/*
+	 * Initialize DDR supply (regulator or STPMIC) in BL31 only when
+	 * required to avoid concurrent access with normal world or OP-TEE
+	 */
+	fixed_regulator_register();
+	initialize_pmic();
+	/* DDR and device tree is no more accessible when clk is deactivated */
+	ddr_sub_system_clk_off();
+	ddr_power_off(fdt, node);
+
+	return 0;
+}
+#endif /* IMAGE_BL31 */
-- 
2.53.0

